// @tamagui/react-native-svg/dist/esm/index.native.js
import { createElement, forwardRef } from "react";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
var Circle = forwardRef(function(p, ref) {
  return createElement("circle", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), ClipPath = forwardRef(function(p, ref) {
  return createElement("clipPath", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Defs = forwardRef(function(p, ref) {
  return createElement("defs", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Ellipse = forwardRef(function(p, ref) {
  return createElement("ellipse", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), ForeignObject = forwardRef(function(p, ref) {
  return createElement("foreignObject", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), G = forwardRef(function(p, ref) {
  return createElement("g", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Image = forwardRef(function(p, ref) {
  return createElement("image", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Line = forwardRef(function(p, ref) {
  return createElement("line", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), LinearGradient = forwardRef(function(p, ref) {
  return createElement("linearGradient", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Marker = forwardRef(function(p, ref) {
  return createElement("marker", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Mask = forwardRef(function(p, ref) {
  return createElement("mask", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Path = forwardRef(function(p, ref) {
  return createElement("path", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Pattern = forwardRef(function(p, ref) {
  return createElement("pattern", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Polygon = forwardRef(function(p, ref) {
  return createElement("polygon", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Polyline = forwardRef(function(p, ref) {
  return createElement("polyline", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), RadialGradient = forwardRef(function(p, ref) {
  return createElement("radialGradient", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Rect = forwardRef(function(p, ref) {
  return createElement("rect", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Shape = forwardRef(function(p, ref) {
  return createElement("shape", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Stop = forwardRef(function(p, ref) {
  return createElement("stop", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Svg = forwardRef(function(p, ref) {
  return createElement("svg", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Symbol = forwardRef(function(p, ref) {
  return createElement("symbol", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Text = forwardRef(function(p, ref) {
  return createElement("text", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), TextPath = forwardRef(function(p, ref) {
  return createElement("clipPath", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), TSpan = forwardRef(function(p, ref) {
  return createElement("tSpan", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Use = forwardRef(function(p, ref) {
  return createElement("use", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), src_default = Svg;
export {
  Circle,
  ClipPath,
  Defs,
  Ellipse,
  ForeignObject,
  G,
  Image,
  Line,
  LinearGradient,
  Marker,
  Mask,
  Path,
  Pattern,
  Polygon,
  Polyline,
  RadialGradient,
  Rect,
  Shape,
  Stop,
  Svg,
  Symbol,
  TSpan,
  Text,
  TextPath,
  Use,
  src_default as default
};
