// @tamagui/react-native-svg/dist/cjs/index.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var src_exports = {};
__export(src_exports, {
  Circle: () => Circle,
  ClipPath: () => ClipPath,
  Defs: () => Defs,
  Ellipse: () => Ellipse,
  ForeignObject: () => ForeignObject,
  G: () => G,
  Image: () => Image,
  Line: () => Line,
  LinearGradient: () => LinearGradient,
  Marker: () => Marker,
  Mask: () => Mask,
  Path: () => Path,
  Pattern: () => Pattern,
  Polygon: () => Polygon,
  Polyline: () => Polyline,
  RadialGradient: () => RadialGradient,
  Rect: () => Rect,
  Shape: () => Shape,
  Stop: () => Stop,
  Svg: () => Svg,
  Symbol: () => Symbol2,
  TSpan: () => TSpan,
  Text: () => Text,
  TextPath: () => TextPath,
  Use: () => Use,
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var import_react = require("react");
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
var Circle = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("circle", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), ClipPath = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("clipPath", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Defs = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("defs", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Ellipse = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("ellipse", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), ForeignObject = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("foreignObject", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), G = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("g", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Image = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("image", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Line = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("line", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), LinearGradient = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("linearGradient", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Marker = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("marker", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Mask = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("mask", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Path = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("path", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Pattern = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("pattern", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Polygon = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("polygon", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Polyline = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("polyline", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), RadialGradient = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("radialGradient", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Rect = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("rect", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Shape = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("shape", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Stop = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("stop", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Svg = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("svg", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Symbol2 = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("symbol", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Text = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("text", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), TextPath = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("clipPath", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), TSpan = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("tSpan", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), Use = (0, import_react.forwardRef)(function(p, ref) {
  return (0, import_react.createElement)("use", _object_spread_props(_object_spread({}, p), {
    ref
  }));
}), src_default = Svg;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Circle,
  ClipPath,
  Defs,
  Ellipse,
  ForeignObject,
  G,
  Image,
  Line,
  LinearGradient,
  Marker,
  Mask,
  Path,
  Pattern,
  Polygon,
  Polyline,
  RadialGradient,
  Rect,
  Shape,
  Stop,
  Svg,
  Symbol,
  TSpan,
  Text,
  TextPath,
  Use
});
