// @tamagui/web/dist/esm/html.js
import React from "react";
import { jsx } from "react/jsx-runtime";
function createHTML(TagName) {
  const component = React.forwardRef(
    (props, forwardedRef) => /* @__PURE__ */ jsx(TagName, { ...props, ref: forwardedRef })
  );
  return component.staticConfig = {
    componentName: TagName,
    Component: component,
    isText: !0,
    acceptsClassName: !0
  }, component.name = TagName, component.displayName = TagName, component;
}
const a = createHTML("a"), abbr = createHTML("abbr"), address = createHTML("address"), area = createHTML("area"), article = createHTML("article"), aside = createHTML("aside"), audio = createHTML("audio"), b = createHTML("b"), base = createHTML("base"), bdi = createHTML("bdi"), bdo = createHTML("bdo"), big = createHTML("big"), blockquote = createHTML("blockquote"), body = createHTML("body"), br = createHTML("br"), button = createHTML("button"), canvas = createHTML("canvas"), caption = createHTML("caption"), center = createHTML("center"), cite = createHTML("cite"), code = createHTML("code"), col = createHTML("col"), colgroup = createHTML("colgroup"), data = createHTML("data"), datalist = createHTML("datalist"), dd = createHTML("dd"), del = createHTML("del"), details = createHTML("details"), dfn = createHTML("dfn"), dialog = createHTML("dialog"), div = createHTML("div"), dl = createHTML("dl"), dt = createHTML("dt"), em = createHTML("em"), embed = createHTML("embed"), fieldset = createHTML("fieldset"), figcaption = createHTML("figcaption"), figure = createHTML("figure"), footer = createHTML("footer"), form = createHTML("form"), h1 = createHTML("h1"), h2 = createHTML("h2"), h3 = createHTML("h3"), h4 = createHTML("h4"), h5 = createHTML("h5"), h6 = createHTML("h6"), head = createHTML("head"), header = createHTML("header"), hgroup = createHTML("hgroup"), hr = createHTML("hr"), html = createHTML("html"), i = createHTML("i"), iframe = createHTML("iframe"), img = createHTML("img"), input = createHTML("input"), ins = createHTML("ins"), kbd = createHTML("kbd"), keygen = createHTML("keygen"), label = createHTML("label"), legend = createHTML("legend"), li = createHTML("li"), link = createHTML("link"), main = createHTML("main"), map = createHTML("map"), mark = createHTML("mark"), menu = createHTML("menu"), menuitem = createHTML("menuitem"), meta = createHTML("meta"), meter = createHTML("meter"), nav = createHTML("nav"), noindex = createHTML("noindex"), noscript = createHTML("noscript"), object = createHTML("object"), ol = createHTML("ol"), optgroup = createHTML("optgroup"), option = createHTML("option"), output = createHTML("output"), p = createHTML("p"), param = createHTML("param"), picture = createHTML("picture"), pre = createHTML("pre"), progress = createHTML("progress"), q = createHTML("q"), rp = createHTML("rp"), rt = createHTML("rt"), ruby = createHTML("ruby"), s = createHTML("s"), samp = createHTML("samp"), search = createHTML("search"), slot = createHTML("slot"), script = createHTML("script"), section = createHTML("section"), select = createHTML("select"), small = createHTML("small"), source = createHTML("source"), span = createHTML("span"), strong = createHTML("strong"), style = createHTML("style"), sub = createHTML("sub"), summary = createHTML("summary"), sup = createHTML("sup"), table = createHTML("table"), template = createHTML("template"), tbody = createHTML("tbody"), td = createHTML("td"), textarea = createHTML("textarea"), tfoot = createHTML("tfoot"), th = createHTML("th"), thead = createHTML("thead"), time = createHTML("time"), title = createHTML("title"), tr = createHTML("tr"), track = createHTML("track"), u = createHTML("u"), ul = createHTML("ul"), var_ = createHTML("var");
const video = createHTML("video"), wbr = createHTML("wbr"), webview = createHTML("webview"), svg = createHTML("svg"), animate = createHTML("animate"), animateMotion = createHTML("animateMotion"), animateTransform = createHTML("animateTransform"), circle = createHTML("circle"), clipPath = createHTML("clipPath"), defs = createHTML("defs"), desc = createHTML("desc"), ellipse = createHTML("ellipse"), feBlend = createHTML("feBlend"), feColorMatrix = createHTML("feColorMatrix"), feComponentTransfer = createHTML("feComponentTransfer"), feComposite = createHTML("feComposite"), feConvolveMatrix = createHTML("feConvolveMatrix"), feDiffuseLighting = createHTML("feDiffuseLighting"), feDisplacementMap = createHTML("feDisplacementMap"), feDistantLight = createHTML("feDistantLight"), feDropShadow = createHTML("feDropShadow"), feFlood = createHTML("feFlood"), feFuncA = createHTML("feFuncA"), feFuncB = createHTML("feFuncB"), feFuncG = createHTML("feFuncG"), feFuncR = createHTML("feFuncR"), feGaussianBlur = createHTML("feGaussianBlur"), feImage = createHTML("feImage"), feMerge = createHTML("feMerge"), feMergeNode = createHTML("feMergeNode"), feMorphology = createHTML("feMorphology"), feOffset = createHTML("feOffset"), fePointLight = createHTML("fePointLight"), feSpecularLighting = createHTML("feSpecularLighting"), feSpotLight = createHTML("feSpotLight"), feTile = createHTML("feTile"), feTurbulence = createHTML("feTurbulence"), filter = createHTML("filter"), foreignObject = createHTML("foreignObject"), g = createHTML("g"), image = createHTML("image"), line = createHTML("line"), linearGradient = createHTML("linearGradient"), marker = createHTML("marker"), mask = createHTML("mask"), metadata = createHTML("metadata"), mpath = createHTML("mpath"), path = createHTML("path"), pattern = createHTML("pattern"), polygon = createHTML("polygon"), polyline = createHTML("polyline"), radialGradient = createHTML("radialGradient"), rect = createHTML("rect"), stop = createHTML("stop"), switch_ = createHTML("switch");
const symbol = createHTML("symbol"), text = createHTML("text"), textPath = createHTML("textPath"), tspan = createHTML("tspan"), use = createHTML("use"), view = createHTML("view");
export {
  a,
  abbr,
  address,
  animate,
  animateMotion,
  animateTransform,
  area,
  article,
  aside,
  audio,
  b,
  base,
  bdi,
  bdo,
  big,
  blockquote,
  body,
  br,
  button,
  canvas,
  caption,
  center,
  circle,
  cite,
  clipPath,
  code,
  col,
  colgroup,
  data,
  datalist,
  dd,
  defs,
  del,
  desc,
  details,
  dfn,
  dialog,
  div,
  dl,
  dt,
  ellipse,
  em,
  embed,
  feBlend,
  feColorMatrix,
  feComponentTransfer,
  feComposite,
  feConvolveMatrix,
  feDiffuseLighting,
  feDisplacementMap,
  feDistantLight,
  feDropShadow,
  feFlood,
  feFuncA,
  feFuncB,
  feFuncG,
  feFuncR,
  feGaussianBlur,
  feImage,
  feMerge,
  feMergeNode,
  feMorphology,
  feOffset,
  fePointLight,
  feSpecularLighting,
  feSpotLight,
  feTile,
  feTurbulence,
  fieldset,
  figcaption,
  figure,
  filter,
  footer,
  foreignObject,
  form,
  g,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6,
  head,
  header,
  hgroup,
  hr,
  html,
  i,
  iframe,
  image,
  img,
  input,
  ins,
  kbd,
  keygen,
  label,
  legend,
  li,
  line,
  linearGradient,
  link,
  main,
  map,
  mark,
  marker,
  mask,
  menu,
  menuitem,
  meta,
  metadata,
  meter,
  mpath,
  nav,
  noindex,
  noscript,
  object,
  ol,
  optgroup,
  option,
  output,
  p,
  param,
  path,
  pattern,
  picture,
  polygon,
  polyline,
  pre,
  progress,
  q,
  radialGradient,
  rect,
  rp,
  rt,
  ruby,
  s,
  samp,
  script,
  search,
  section,
  select,
  slot,
  small,
  source,
  span,
  stop,
  strong,
  style,
  sub,
  summary,
  sup,
  svg,
  switch_ as switch,
  symbol,
  table,
  tbody,
  td,
  template,
  text,
  textPath,
  textarea,
  tfoot,
  th,
  thead,
  time,
  title,
  tr,
  track,
  tspan,
  u,
  ul,
  use,
  var_ as var,
  video,
  view,
  wbr,
  webview
};
