// @tamagui/toast/dist/jsx/Toast.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { styled, useEvent } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { ThemeableStack } from "@tamagui/stacks";
import { SizableText } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { TOAST_NAME } from "./constants";
import { ToastAnnounceExclude } from "./ToastAnnounce";
import { useToast, useToastController, useToastState } from "./ToastImperative";
import { ToastImpl, ToastImplFrame, useToastInteractiveContext } from "./ToastImpl";
import { ToastProvider } from "./ToastProvider";
import { ToastViewport } from "./ToastViewport";
function _array_like_to_array(arr, len) {
  (len == null || len > arr.length) && (len = arr.length);
  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];
  return arr2;
}
function _array_with_holes(arr) {
  if (Array.isArray(arr)) return arr;
}
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _iterable_to_array_limit(arr, i) {
  var _i = arr == null ? null : typeof Symbol < "u" && arr[Symbol.iterator] || arr["@@iterator"];
  if (_i != null) {
    var _arr = [], _n = !0, _d = !1, _s, _e;
    try {
      for (_i = _i.call(arr); !(_n = (_s = _i.next()).done) && (_arr.push(_s.value), !(i && _arr.length === i)); _n = !0)
        ;
    } catch (err) {
      _d = !0, _e = err;
    } finally {
      try {
        !_n && _i.return != null && _i.return();
      } finally {
        if (_d) throw _e;
      }
    }
    return _arr;
  }
}
function _non_iterable_rest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null) return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null) return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
function _sliced_to_array(arr, i) {
  return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
  if (o) {
    if (typeof o == "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor && (n = o.constructor.name), n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
  }
}
var TITLE_NAME = "ToastTitle", ToastTitle = styled(SizableText, {
  name: TITLE_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color",
        size: "$4"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
ToastTitle.displayName = TITLE_NAME;
var DESCRIPTION_NAME = "ToastDescription", ToastDescription = styled(SizableText, {
  name: DESCRIPTION_NAME,
  variants: {
    unstyled: {
      false: {
        color: "$color11",
        size: "$1"
      }
    }
  },
  defaultVariants: {
    unstyled: process.env.TAMAGUI_HEADLESS === "1"
  }
});
ToastDescription.displayName = DESCRIPTION_NAME;
var ACTION_NAME = "ToastAction", ToastAction = /* @__PURE__ */ React.forwardRef(function(props, forwardedRef) {
  var altText2 = props.altText, actionProps = _object_without_properties(props, [
    "altText"
  ]);
  return altText2 ? /* @__PURE__ */ _jsx(ToastAnnounceExclude, {
    altText: altText2,
    asChild: !0,
    children: /* @__PURE__ */ _jsx(ToastClose, _object_spread_props(_object_spread({}, actionProps), {
      ref: forwardedRef
    }))
  }) : null;
});
ToastAction.propTypes = {
  altText: function(props) {
    return props.altText ? null : new Error("Missing prop `altText` expected on `".concat(ACTION_NAME, "`"));
  }
};
ToastAction.displayName = ACTION_NAME;
var CLOSE_NAME = "ToastClose", ToastCloseFrame = styled(ThemeableStack, {
  name: CLOSE_NAME,
  tag: "button"
}), ToastClose = /* @__PURE__ */ React.forwardRef(function(props, forwardedRef) {
  var __scopeToast = props.__scopeToast, closeProps = _object_without_properties(props, [
    "__scopeToast"
  ]), interactiveContext = useToastInteractiveContext(__scopeToast);
  return /* @__PURE__ */ _jsx(ToastAnnounceExclude, {
    asChild: !0,
    children: /* @__PURE__ */ _jsx(ToastCloseFrame, _object_spread_props(_object_spread({
      accessibilityLabel: "Dialog Close"
    }, closeProps), {
      ref: forwardedRef,
      onPress: composeEventHandlers(props.onPress, interactiveContext.onClose)
    }))
  });
});
ToastClose.displayName = CLOSE_NAME;
var ToastComponent = ToastImplFrame.styleable(function(props, forwardedRef) {
  var forceMount = props.forceMount, openProp = props.open, defaultOpen = props.defaultOpen, onOpenChange = props.onOpenChange, toastProps = _object_without_properties(props, [
    "forceMount",
    "open",
    "defaultOpen",
    "onOpenChange"
  ]), _useControllableState = _sliced_to_array(useControllableState({
    prop: openProp,
    defaultProp: defaultOpen ?? !0,
    onChange: onOpenChange,
    strategy: "most-recent-wins"
  }), 2), open = _useControllableState[0], setOpen = _useControllableState[1], id = React.useId(), onPause = useEvent(props.onPause), onResume = useEvent(props.onResume), shouldShow = forceMount || open;
  return shouldShow ? /* @__PURE__ */ _jsx(ToastImpl, _object_spread_props(_object_spread({
    id,
    open
  }, toastProps), {
    ref: forwardedRef,
    onClose: function() {
      return setOpen(!1);
    },
    onPause,
    onResume,
    onSwipeEnd: composeEventHandlers(props.onSwipeEnd, function(event) {
      setOpen(!1);
    })
  })) : null;
});
ToastComponent.displayName = TOAST_NAME;
var Toast = withStaticProperties(ToastComponent, {
  Title: ToastTitle,
  Description: ToastDescription,
  Action: ToastAction,
  Close: ToastClose
});
export {
  Toast,
  ToastProvider,
  ToastViewport,
  useToast,
  useToastController,
  useToastState
};
